tls/extensions/s2n_client_key_share.c

```diff
@@ -111,24 +111,30 @@ static int s2n_generate_pq_hybrid_key_share(struct s2n_stuffer *out, struct s2n_
     const struct s2n_kem_group *kem_group = kem_group_params->kem_group;
     POSIX_ENSURE_REF(kem_group);
 
-    /* The structure of the PQ share is:
-     *    IANA ID (2 bytes)
-     * || total share size (2 bytes)
-     * || size of ECC key share (2 bytes)
-     * || ECC key share (variable bytes)
-     * || size of PQ key share (2 bytes)
-     * || PQ key share (variable bytes) */
     POSIX_GUARD(s2n_stuffer_write_uint16(out, kem_group->iana_id));
 
     struct s2n_stuffer_reservation total_share_size = { 0 };
     POSIX_GUARD(s2n_stuffer_reserve_uint16(out, &total_share_size));
 
     struct s2n_ecc_evp_params *ecc_params = &kem_group_params->ecc_params;
     ecc_params->negotiated_curve = kem_group->curve;
-    POSIX_GUARD_RESULT(s2n_ecdhe_send_public_key(ecc_params, out));
 
     struct s2n_kem_params *kem_params = &kem_group_params->kem_params;
     kem_params->kem = kem_group->kem;
+
+    POSIX_ENSURE_REF(ecc_params);
+    POSIX_ENSURE_REF(ecc_params->negotiated_curve);
+
+    if (kem_params->len_prefixed) {
+        POSIX_GUARD(s2n_stuffer_write_uint16(out, ecc_params->negotiated_curve->share_size));
+    }
+
+    if (ecc_params->evp_pkey == NULL) {
+        POSIX_GUARD(s2n_ecc_evp_generate_ephemeral_key(ecc_params));
+    }
+
+    POSIX_GUARD(s2n_ecc_evp_write_params_point(ecc_params, out));
```

Comment:
Except for the kem_params->len_prefixed check, this is a duplicate of s2n_ecdhe_send_public_key :\ Maybe we should just move writing the length out of s2n_ecdhe_send_public_key? Or have a "include length" flag input? Or have a common s2n_ecdhe_send_public_key_value method called by s2n_ecdhe_send_public_key and this method?

But on the other hand, it's only a few lines of logic, so maybe it's not worth it?

Reply:
Yeah, I went back and forth on whether to update `s2n_ecdhe_send_public_key()` or copy it's implementation here. I swapped back to updating `s2n_ecdhe_send_public_key()` to take a new `len_prefixed` variable, and update it's callers.